src/java/org/apache/cassandra/db/DeletionTime.java

```diff
@@ -38,14 +41,15 @@ public class DeletionTime implements Comparable<DeletionTime>, IMeasurableMemory
     /**
      * A special DeletionTime that signifies that there is no top-level (row) tombstone.
      */
-    public static final DeletionTime LIVE = new DeletionTime(Long.MIN_VALUE, Integer.MAX_VALUE);
+    public static final DeletionTime LIVE = new DeletionTime(Long.MIN_VALUE, Long.MAX_VALUE);
 
     public static final Serializer serializer = new Serializer();
+    public static final Serializer legacySerializer = new LegacySerializer();
 
     private final long markedForDeleteAt;
-    private final int localDeletionTime;
+    private final long localDeletionTime;
```

Comment:
Why do we make this `long` when we only ever use the lower 32 bits?

Reply:
Converting uint<-> long should have some perf impact hence having it readily available as a long it's nice. But given I do indeed store an int for [Cells](https://github.com/apache/cassandra/pull/1891/files/449129ca26b83efed084471d822a52f81cb3a0c6#diff-bdff7d68605bca83e303d541e5c301d289c2b0fd54a9b04e2185ffc5d2ca8770R42) which is the same optimization i.e., I should perf test moving this to an int and see if it has an impact.

Reply:
It really shouldn't. The cast is free, you just load into a suitably sized register. The truncation will have a single cycle latency (and greater than one-per-cycle throughput). If you can measure the impact of this in a real C* trace, I will eat my hat. If you can even get it to show up in a microbenchmark that does _anything_ real (compaction, memtable read or anything really) I will be very impressed. It's going to be 100% swallowed up by memory latency, data dependencies in the pipeline, etc. 

However, the wasted memory will definitely have an impact. For many JVMs this will require an extra 8-bytes of memory. 

So this is a pretty simple tradeoff, and I would not personally waste my time benchmarking it.

Reply:
+1 to using `int`, the same way we do it in {{ArrayCell}} and {{BufferCell}}.

Reply:
Testing this the problem is that when we're scrubbing legacy sstables with overflowed ldts we need to support reading negative ldts. Moving this to an uint is not possible then which is why probably this wasn't moved in the first place. I leave this one pending.

Reply:
Extend DeletionTime for those cases, so that it interprets as signed?

Reply:
Yep, the problem with `DeletionTime#validate` can be addressed with a class extending `DeletionTime`, probably `InvalidDeletionTime`, or `OverflownDeletionTime`. There could be a factory method such as:
```java
public static DeletionTime DeletionTime(long markedForDeleteAt, long localDeletionTime)
{
    return localDeletionTime >= 0 
           ? new DeletionTime(markedForDeleteAt, localDeletionTime) 
           : new OverflownDeletionTime(markedForDeleteAt, localDeletionTime);
}
```
Then `DeletionTime` and `InvalidDeletionTime` can have implementations of `validate()` that just return `true` or `false`.

Reply:
Added in the latest commit.